docs/proposals/kubemark.md

```diff
@@ -0,0 +1,97 @@
+
+## Kubemark architecture overview
+The high-level idea behind Kubemark is to write an artificial “NodeHolograms” that will be able to simulate a behavior of real Kubelet and KubeProxy in a single, lightweight binary. Holo-components will need to correctly respond to Controllers (via API server), and preferably, in the fullness of time, be able to ‘replay’ real traffic (this is out of scope for initial version). In the initial version we only want Holograms to be able to fool Master components and put some configurable (in what way TBD) load on them.
+
+When we have Holograms ready, we’ll be able to test performance of Master Components by creating a real Master machine, with API server, Controllers, etcd and whatnot, and create number of HoloNodes that will register to the running Master. 
+
+For the initial version we will teach Master components to use port number to identify Kubelet/KubeProxy. This will allow running those components on non-default ports, and in the same time will allow to run multiple Holograms on a single machine. During setup we will generate credentials for cluster communication and pass them to HoloKubelet/HoloProxy to use. Master will treat all HoloNodes as normal ones.
+
+## Things to simulate
+In real Kubernetes on a single Node we run two daemons that communicate with Master in some way: Kubelet and KubeProxy. In this paragraph we list what kind of communication we have between those daemons and Master components. First step in creating holograms is understanding the communication between Master and Node components.
+### KubeProxy
+Creates two Watch clients, one for Services, one for Endpoints (it’s done in pkg/proxy/config/api.go, because why not). KubeProxy also exposes healthz, but it’s not read by the Master (or anything else except supervisord AFAICT). Because this is a passive component (i.e. does not write anything to API server) simulating it is trivial - we just need to create those watches and it will generate exactly the same load as real KubeProxy.
+
+### Kubelet
+More complex of the two. Kubelet watches Services, and Nodes (in separate Watch clients, created in pkg/kubelet/kubelet.go) and Pods in another client (created in pkg/kubelet/config/apiserver.go because it’s the most obvious place to put it). On top of reading data from API server using those 3 Watch clients Kubelet pushes updates of PodStatuses and NodeStatuses using appropriate subresources.
```

Comment:
Did you consider reusing a real Kubelet with some "fake" components?
I started looking into Kubelet and it is quite well structured. Maybe replacing some if its "subcomponents" with fake implementations will solve the problem (and we will have a real Kubelet which is always good).
Looking on the Kubelet class it seems that we would need to replace:

- DockerInterface
- RuntimeCache (FakeRuntimeCache already exists)
- ContainerCommandRunner (FakeRuntimeCache implements it)
- VolumePluginMgr
- NetworkPluginMgr (nooopNetworkPlugin already exists)
- containerGC (should be extremely trivial - one empty method)
- imageManager (should be extremely trivial - one empty method)
- diskSpaceManager
- cloudProvider interface (should be set of empty methods)
- Runtime (FakeRuntime already exists)
- OSInteface (FakeOS already exists)
- mountInterface
- containerManager (should be very simple - no-op)

So maybe this would be very simple and obviously it would be much more accurate (and we avoid diverging HoloKubelet from a real Kubelet in the future).

Any thoughts?

@timothysc @erictune